cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,374 @@
+ */
+#include <cudf/copying.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/scatter.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/detail/utilities/cuda.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {
+
+std::unique_ptr<column> copy_with_nullable(column_view const& original,
+    bool force_mask_allocation, rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+{
+  auto copy = std::make_unique<column>(original, stream, mr);
+  if (force_mask_allocation and not original.nullable()) {
+    // Create an all-valid null mask if original has none
+    auto contents = copy->release();
+    auto bitmask = create_null_mask(original.size(), mask_state::ALL_VALID, stream, mr);
+    size_type null_count = 0;
+    copy = std::make_unique<column>(original.type(), original.size(),
+      std::move(*contents.data), std::move(bitmask), null_count, std::move(contents.children));
+    // TODO use this from PR 3172 instead
+    //auto mask = create_null_mask(target.size(), mask_state::ALL_VALID, stream, mr);
+    //copy->set_null_mask(std::move(mask), null_count);
+  }
+  return copy;
+}
+
+template <typename T>
+thrust::device_vector<T> make_gather_map(column_view const& scatter_map,
+    size_type rows, cudaStream_t stream)
+{
+  // Transform negative indices to index + rows
+  auto scatter_iter = thrust::make_transform_iterator(
+    scatter_map.begin<T>(),
+    index_converter<T>{rows});
+
+  static_assert(std::is_signed<T>::value,
+    "Need different invalid index if unsigned index types are added");
+  auto const invalid_index = static_cast<T>(-1);
+
+  // Convert scatter map to a gather map
+  auto gather_map = thrust::device_vector<T>(rows, invalid_index);
+  thrust::scatter(rmm::exec_policy(stream)->on(stream),
+    thrust::make_counting_iterator<T>(0),
+    thrust::make_counting_iterator<T>(scatter_map.size()),
+    scatter_iter, gather_map.begin());
+
+  return gather_map;
+}
+
+template <typename MapIterator>
+void gather_bitmask(table_view const& source, MapIterator gather_map,
+    std::vector<std::unique_ptr<column>>& target, cudaStream_t stream)
+{
+  auto const device_source = table_device_view::create(source, stream);
+
+  // Make mutable table view from columns
+  auto target_views = std::vector<mutable_column_view>(target.size());
+  std::transform(target.begin(), target.end(), target_views.begin(),
+    [](auto const& col) { return static_cast<mutable_column_view>(*col); });
+  auto target_table = mutable_table_view(target_views);
+  auto device_target = mutable_table_device_view::create(target_table, stream);
+
+  // Compute block size
+  int grid_size, block_size;
+  auto bitmask_kernel = gather_bitmask_kernel<true, decltype(gather_map)>;
+  CUDA_TRY(cudaOccupancyMaxPotentialBlockSize(&grid_size, &block_size, bitmask_kernel));
```

Comment:
Since we don't have benchmarks to tune this, why not just hard code it to something like 256 like we do everywhere else? The max occupancy block size doesn't guarantee the fastest execution...

Reply:
Done